File: src/scoring.ts
import * as fs from 'fs';
import * as path from 'path';

export interface ScoreAgentConfig {
  key: string;
  name: string;
  connectorId: string;
  responseCsv: string;
}

interface EvalItem {
  id?: string;
  prompt?: string;
  expected_answer?: string;
  expectedAnswer?: string;
  assertions?: Array<{ value?: string; wholeWord?: boolean }>;
  supporting_facts?: string[];
  supportingFacts?: string[];
  category?: string;
  difficulty?: string;
}

const NO_RESULT_PATTERNS = [
  'no matching',
  'not found',
  'unable to return',
  'was not located',
  'no records found',
  'could not find',
];

export function scoreResponseSet(evalgenJson: string, agents: ScoreAgentConfig[], outputDir: string): void {
  const evalItems = readEvalItems(evalgenJson);
  const payload: Record<string, unknown> = {
    methodology: {
      deterministic_grounding_score: '80% EvalGen must-contain assertion coverage + 20% supporting-fact value coverage',
      semantic_quality_score: 'Local semantic quality fallback: token-overlap F1 between expected and actual answer, separate from deterministic grounding.',
      matching: 'Case-insensitive, Unicode-normalized, punctuation/spacing-tolerant matching; whole-word assertions enforce token boundaries.',
    },
    agents: {},
  };
  const agentPayload = payload.agents as Record<string, unknown>;

  for (const agent of agents) {
    const rows = readCsv(agent.responseCsv);
    const scored = evalItems.map((item, index) => scoreItem(item, rows[index] || {}, index + 1));
    agentPayload[agent.key] = {
      summary: summarize(agent, scored),
      items: scored,
      category_summary: summarizeCategories(scored),
    };
  }

  fs.mkdirSync(outputDir, { recursive: true });
  fs.writeFileSync(path.join(outputDir, 'agent-response-scores.json'), `${JSON.stringify(payload, null, 2)}\n`, 'utf-8');
  fs.writeFileSync(path.join(outputDir, 'agent-response-scores.md'), renderMarkdown(payload, agents), 'utf-8');
}

function scoreItem(item: EvalItem, row: Record<string, string>, index: number): Record<string, unknown> {
  const response = row.actual_answer || '';
  const assertions = item.assertions || [];
  const assertionResults = assertions.map((assertion) => {
    const value = assertion.value || '';
    return { value, passed: containsValue(response, value, !!assertion.wholeWord), whole_word: !!assertion.wholeWord };
  });
  const factResults = normalizeFacts(item.supporting_facts || item.supportingFacts || []).map((fact) => {
    const [, value] = parseFact(fact);
    return { value, passed: containsValue(response, value, isShortCode(value)) };
  }).filter((fact) => fact.value);

  const assertionsTotal = assertionResults.length;
  const assertionsPassed = assertionResults.filter((result) => result.passed).length;
  const factsTotal = factResults.length;
  const factsPassed = factResults.filter((result) => result.passed).length;
  const expected = item.expected_answer || item.expectedAnswer || row.expected_answer || '';
  const expectedNoResult = hasNoResultLanguage(expected);
  const responseNoResult = hasNoResultLanguage(response);

  let assertionScore = 0;
  let factScore = 0;
  let deterministicScore = 0;
  let status: 'pass' | 'partial' | 'fail' = 'fail';
  if (assertionsTotal > 0) {
    assertionScore = assertionsPassed / assertionsTotal;
    factScore = factsTotal > 0 ? factsPassed / factsTotal : assertionScore;
    deterministicScore = (0.8 * assertionScore) + (0.2 * factScore);
    status = assertionsPassed === assertionsTotal ? 'pass' : assertionsPassed > 0 ? 'partial' : 'fail';
  } else if (factsTotal > 0) {
    assertionScore = factsPassed / factsTotal;
    factScore = assertionScore;
    deterministicScore = factScore;
    status = factsPassed === factsTotal ? 'pass' : factsPassed > 0 ? 'partial' : 'fail';
  } else if (expectedNoResult) {
    assertionScore = responseNoResult ? 1 : 0;
    factScore = assertionScore;
    deterministicScore = assertionScore;
    status = responseNoResult ? 'pass' : 'fail';
  }

  return {
    index,
    id: item.id || '',
    category: item.category || '',
    difficulty: item.difficulty || '',
    prompt: item.prompt || row.prompt || '',
    expected_answer: expected,
    actual_answer: response,
    assertions: assertionResults,
    supporting_facts: factResults,
    assertions_passed: assertionsPassed,
    assertions_total: assertionsTotal,
    facts_passed: factsPassed,
    facts_total: factsTotal,
    deterministic_grounding_score: roundPct(deterministicScore),
    semantic_quality_score: roundPct(localSemanticQuality(expected, response)),
    semantic_quality_provider: 'local-token-overlap',
    assertion_score: roundPct(assertionScore),
    fact_score: roundPct(factScore),
    status,
    has_citation: hasCitation(response),
    has_no_result_language: responseNoResult,
    expected_no_result: expectedNoResult,
    failed_checks: assertionResults.filter((result) => !result.passed).map((result) => result.value),
  };
}

function summarize(agent: ScoreAgentConfig, scores: Array<Record<string, unknown>>): Record<string, unknown> {
  const deterministic = scores.map((item) => Number(item.deterministic_grounding_score || 0));
  const semantic = scores.map((item) => Number(item.semantic_quality_score || 0));
  const totalAssertions = sum(scores, 'assertions_total');
  const passedAssertions = sum(scores, 'assertions_passed');
  const totalFacts = sum(scores, 'facts_total');
  const passedFacts = sum(scores, 'facts_passed');
  return {
    agent: { key: agent.key, name: agent.name, connector_id: agent.connectorId },
    prompt_count: scores.length,
    average_deterministic_grounding_score: average(deterministic),
    average_semantic_quality_score: average(semantic),
    assertion_pass_rate: totalAssertions ? roundPct(passedAssertions / totalAssertions) : 100,
    assertions_passed: passedAssertions,
    assertions_total: totalAssertions,
    fact_pass_rate: totalFacts ? roundPct(passedFacts / totalFacts) : 100,
    facts_passed: passedFacts,
    facts_total: totalFacts,
    pass_count: scores.filter((item) => item.status === 'pass').length,
    partial_count: scores.filter((item) => item.status === 'partial').length,
    fail_count: scores.filter((item) => item.status === 'fail').length,
    citation_count: scores.filter((item) => item.has_citation).length,
  };
}

function summarizeCategories(scores: Array<Record<string, unknown>>): Record<string, unknown> {
  const categories = [...new Set(scores.map((item) => String(item.category || 'uncategorized')))].sort();
  const out: Record<string, unknown> = {};
  for (const category of categories) {
    const items = scores.filter((item) => String(item.category || 'uncategorized') === category);
    out[category] = {
      count: items.length,
      average_deterministic_grounding_score: average(items.map((item) => Number(item.deterministic_grounding_score || 0))),
      average_semantic_quality_score: average(items.map((item) => Number(item.semantic_quality_score || 0))),
    };
  }
  return out;
}

function renderMarkdown(payload: Record<string, unknown>, agents: ScoreAgentConfig[]): string {
  const agentsPayload = payload.agents as Record<string, { summary: Record<string, unknown> }>;
  const lines = [
    '# Agent Response Scoring',
    '',
    'Scores include deterministic grounding and a separate semantic quality score.',
    '',
    '| Agent | Connector | Avg grounding | Avg semantic quality | Assertions passed | Fact pass rate | Pass | Partial | Fail |',
    '|---|---|---:|---:|---:|---:|---:|---:|---:|',
  ];
  for (const agent of agents) {
    const summary = agentsPayload[agent.key]?.summary;
    if (!summary) continue;
    lines.push(
      `| ${agent.name} | \`${agent.connectorId}\` | ${summary.average_deterministic_grounding_score}% | ` +
      `${summary.average_semantic_quality_score}% | ${summary.assertions_passed}/${summary.assertions_total} | ` +
      `${summary.fact_pass_rate}% | ${summary.pass_count} | ${summary.partial_count} | ${summary.fail_count} |`,
    );
  }
  return `${lines.join('\n')}\n`;
}

function readEvalItems(filePath: string): EvalItem[] {
  const parsed = JSON.parse(fs.readFileSync(filePath, 'utf-8')) as { items?: EvalItem[] };
  if (!Array.isArray(parsed.items)) throw new Error(`EvalGen JSON has no items array: ${filePath}`);
  return parsed.items;
}

function readCsv(filePath: string): Array<Record<string, string>> {
  const rows = parseCsv(fs.readFileSync(filePath, 'utf-8'));
  const header = rows[0] || [];
  return rows.slice(1).map((row) => {
    const out: Record<string, string> = {};
    for (let i = 0; i < header.length; i++) out[header[i]] = row[i] || '';
    return out;
  });
}

function parseCsv(content: string): string[][] {
  const rows: string[][] = [];
  let row: string[] = [];
  let cell = '';
  let inQuotes = false;
  const text = stripBom(content);
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (ch === '"') {
      if (inQuotes && text[i + 1] === '"') {
        cell += '"';
        i++;
      } else {
        inQuotes = !inQuotes;
      }
    } else if (ch === ',' && !inQuotes) {
      row.push(cell);
      cell = '';
    } else if ((ch === '\n' || ch === '\r') && !inQuotes) {
      if (ch === '\r' && text[i + 1] === '\n') i++;
      row.push(cell);
      rows.push(row);
      row = [];
      cell = '';
    } else {
      cell += ch;
    }
  }
  if (cell || row.length) {
    row.push(cell);
    rows.push(row);
  }
  return rows.filter((cells) => cells.some((value) => value.trim()));
}

function foldText(value: string): string {
  return (value || '').normalize('NFKD').replace(/\p{Diacritic}/gu, '').toLowerCase();
}

function compact(value: string): string {
  return foldText(value).replace(/[^a-z0-9]+/g, '');
}

function containsValue(response: string, expected: string, wholeWord: boolean): boolean {
  if (!expected) return true;
  const foldedResponse = foldText(response);
  const foldedExpected = foldText(expected);
  if (wholeWord || isShortCode(expected)) {
    const pattern = new RegExp(`(?<![a-z0-9])${escapeRegex(foldedExpected)}(?![a-z0-9])`);
    if (pattern.test(foldedResponse)) return true;
  }
  return foldedResponse.includes(foldedExpected) || compact(response).includes(compact(expected));
}

function localSemanticQuality(expected: string, actual: string): number {
  const expectedTokens = tokenSet(expected);
  const actualTokens = tokenSet(actual);
  if (expectedTokens.size === 0) return actualTokens.size === 0 ? 1 : 0;
  if (actualTokens.size === 0) return 0;
  let overlap = 0;
  for (const token of expectedTokens) if (actualTokens.has(token)) overlap++;
  const precision = overlap / actualTokens.size;
  const recall = overlap / expectedTokens.size;
  return precision + recall === 0 ? 0 : (2 * precision * recall) / (precision + recall);
}

function tokenSet(value: string): Set<string> {
  return new Set(foldText(value).split(/[^a-z0-9.]+/).filter((token) => token.length > 1));
}

function isShortCode(value: string): boolean {
  const stripped = (value || '').replace(/[^A-Za-z0-9]/g, '');
  return stripped.length > 1 && stripped.length <= 3 && stripped.toUpperCase() === stripped;
}

function hasNoResultLanguage(value: string): boolean {
  const folded = foldText(value);
  return NO_RESULT_PATTERNS.some((pattern) => folded.includes(pattern));
}

function hasCitation(value: string): boolean {
  return value.includes('\ue200cite') || value.includes('cite') || /\[\^\d+\^\]/.test(value);
}

function normalizeFacts(facts: string[]): string[] {
  return facts.filter((fact) => typeof fact === 'string' && fact.trim());
}

function parseFact(fact: string): [string, string] {
  const index = fact.indexOf('=');
  return index < 0 ? ['', fact.trim()] : [fact.slice(0, index).trim(), fact.slice(index + 1).trim()];
}

function average(values: number[]): number {
  return values.length ? Math.round((values.reduce((a, b) => a + b, 0) / values.length) * 10) / 10 : 0;
}

function sum(values: Array<Record<string, unknown>>, key: string): number {
  return values.reduce((total, item) => total + Number(item[key] || 0), 0);
}

function roundPct(value: number): number {
  return Math.round(value * 1000) / 10;
}

function escapeRegex(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function stripBom(value: string): string {
  return value.charCodeAt(0) === 0xfeff ? value.slice(1) : value;
}

File: winui3/src/Ccw.Core/Scoring/Scoring.cs
// Port of src/scoring.ts. Step 6 deterministic + local-semantic scoring.
//
// Produces 06-score/agent-response-scores.json + .md from an evalgen JSON
// (item set) plus one or more per-agent response CSVs. The JSON shape is
// the canonical persisted artifact consumed by `compare-jobs` and by the
// UI, so byte-equivalence with the Node output matters.
//
// PARITY NOTES (vs TS):
// * containsValue uses Unicode NFKD + diacritic strip + ASCII regex with
//   lookbehind. .NET supports lookbehind. Diacritic stripping is done by
//   removing Unicode chars whose category is NonSpacingMark after NFKD.
// * roundPct = round(x * 1000) / 10. JS Math.round is half-away-from-zero;
//   we match using MidpointRounding.AwayFromZero.
// * NO_RESULT_PATTERNS list matches src/scoring.ts. Folded match.
// * Local semantic F1: 2*p*r/(p+r), tokens are lowercased NFKD letters,
//   length>1, split on non [a-z0-9.].

using System.Globalization;
using System.IO;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

namespace Ccw.Core.Scoring;

/// <summary>One agent's response CSV plus identifying metadata.</summary>
public sealed record ScoreAgentConfig
{
    public required string Key { get; init; }
    public required string Name { get; init; }
    public required string ConnectorId { get; init; }
    public required string ResponseCsv { get; init; }
}

/// <summary>Top-level entry point. Mirrors TS scoreResponseSet.</summary>
public static class ResponseScorer
{
    private static readonly string[] NoResultPatterns =
    [
        "no matching",
        "not found",
        "unable to return",
        "was not located",
        "no records found",
        "could not find",
    ];

    /// <summary>Reads evalgen.json + per-agent CSVs, writes
    /// <c>agent-response-scores.json</c> + <c>agent-response-scores.md</c>
    /// under <paramref name="outputDir"/>.</summary>
    public static void ScoreResponseSet(string evalgenJson, IReadOnlyList<ScoreAgentConfig> agents, string outputDir)
    {
        var evalItems = ReadEvalItems(evalgenJson);
        var agentsPayload = new Dictionary<string, object>(StringComparer.Ordinal);
        var payload = new Dictionary<string, object>(StringComparer.Ordinal)
        {
            ["methodology"] = new Dictionary<string, string>(StringComparer.Ordinal)
            {
                ["deterministic_grounding_score"] = "80% EvalGen must-contain assertion coverage + 20% supporting-fact value coverage",
                ["semantic_quality_score"] = "Local semantic quality fallback: token-overlap F1 between expected and actual answer, separate from deterministic grounding.",
                ["matching"] = "Case-insensitive, Unicode-normalized, punctuation/spacing-tolerant matching; whole-word assertions enforce token boundaries.",
            },
            ["agents"] = agentsPayload,
        };

        foreach (var agent in agents)
        {
            var rows = ReadCsv(agent.ResponseCsv);
            var scored = new List<Dictionary<string, object?>>(evalItems.Count);
            for (var i = 0; i < evalItems.Count; i++)
            {
                var row = i < rows.Count ? rows[i] : new Dictionary<string, string>(StringComparer.Ordinal);
                scored.Add(ScoreItem(evalItems[i], row, i + 1));
            }

            agentsPayload[agent.Key] = new Dictionary<string, object>(StringComparer.Ordinal)
            {
                ["summary"] = Summarize(agent, scored),
                ["items"] = scored,
                ["category_summary"] = SummarizeCategories(scored),
            };
        }

        Directory.CreateDirectory(outputDir);
        var json = JsonSerializer.Serialize(payload, ScoringJsonOptions);
        File.WriteAllText(Path.Combine(outputDir, "agent-response-scores.json"), json + "\n");
        File.WriteAllText(Path.Combine(outputDir, "agent-response-scores.md"), RenderMarkdown(payload, agents));
    }

    private static readonly JsonSerializerOptions ScoringJsonOptions = new()
    {
        WriteIndented = true,
        NewLine = "\n",
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
        DefaultIgnoreCondition = JsonIgnoreCondition.Never,
    };

    internal sealed class EvalItem
    {
        public string? id { get; set; }
        public string? prompt { get; set; }
        public string? expected_answer { get; set; }
        public string? expectedAnswer { get; set; }
        public List<EvalAssertion>? assertions { get; set; }
        public List<string>? supporting_facts { get; set; }
        public List<string>? supportingFacts { get; set; }
        public string? category { get; set; }
        public string? difficulty { get; set; }
    }

    internal sealed class EvalAssertion
    {
        public string? value { get; set; }
        public bool? wholeWord { get; set; }
    }

    private static List<EvalItem> ReadEvalItems(string filePath)
    {
        using var doc = JsonDocument.Parse(File.ReadAllText(filePath));
        if (!doc.RootElement.TryGetProperty("items", out var items) || items.ValueKind != JsonValueKind.Array)
        {
            throw new InvalidOperationException($"EvalGen JSON has no items array: {filePath}");
        }
        var out_ = new List<EvalItem>(items.GetArrayLength());
        foreach (var item in items.EnumerateArray())
        {
            out_.Add(item.Deserialize<EvalItem>() ?? new EvalItem());
        }
        return out_;
    }

    private static Dictionary<string, object?> ScoreItem(EvalItem item, Dictionary<string, string> row, int index)
    {
        var response = row.GetValueOrDefault("actual_answer") ?? string.Empty;
        var assertions = item.assertions ?? [];
        var assertionResults = new List<Dictionary<string, object?>>(assertions.Count);
        foreach (var a in assertions)
        {
            var value = a.value ?? string.Empty;
            assertionResults.Add(new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["value"] = value,
                ["passed"] = ContainsValue(response, value, a.wholeWord == true),
                ["whole_word"] = a.wholeWord == true,
            });
        }

        var facts = NormalizeFacts(item.supporting_facts ?? item.supportingFacts ?? []);
        var factResults = new List<Dictionary<string, object?>>();
        foreach (var fact in facts)
        {
            var (_, value) = ParseFact(fact);
            if (string.IsNullOrEmpty(value)) continue;
            factResults.Add(new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["value"] = value,
                ["passed"] = ContainsValue(response, value, IsShortCode(value)),
            });
        }

        var assertionsTotal = assertionResults.Count;
        var assertionsPassed = assertionResults.Count(r => (bool)r["passed"]!);
        var factsTotal = factResults.Count;
        var factsPassed = factResults.Count(r => (bool)r["passed"]!);
        var expected = item.expected_answer ?? item.expectedAnswer ?? row.GetValueOrDefault("expected_answer") ?? string.Empty;
        var expectedNoResult = HasNoResultLanguage(expected);
        var responseNoResult = HasNoResultLanguage(response);

        double assertionScore = 0, factScore = 0, deterministicScore = 0;
        string status = "fail";
        if (assertionsTotal > 0)
        {
            assertionScore = (double)assertionsPassed / assertionsTotal;
            factScore = factsTotal > 0 ? (double)factsPassed / factsTotal : assertionScore;
            deterministicScore = (0.8 * assertionScore) + (0.2 * factScore);
            status = assertionsPassed == assertionsTotal ? "pass" : assertionsPassed > 0 ? "partial" : "fail";
        }
        else if (factsTotal > 0)
        {
            assertionScore = (double)factsPassed / factsTotal;
            factScore = assertionScore;
            deterministicScore = factScore;
            status = factsPassed == factsTotal ? "pass" : factsPassed > 0 ? "partial" : "fail";
        }
        else if (expectedNoResult)
        {
            assertionScore = responseNoResult ? 1 : 0;
            factScore = assertionScore;
            deterministicScore = assertionScore;
            status = responseNoResult ? "pass" : "fail";
        }

        return new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["index"] = index,
            ["id"] = item.id ?? string.Empty,
            ["category"] = item.category ?? string.Empty,
            ["difficulty"] = item.difficulty ?? string.Empty,
            ["prompt"] = item.prompt ?? row.GetValueOrDefault("prompt") ?? string.Empty,
            ["expected_answer"] = expected,
            ["actual_answer"] = response,
            ["assertions"] = assertionResults,
            ["supporting_facts"] = factResults,
            ["assertions_passed"] = assertionsPassed,
            ["assertions_total"] = assertionsTotal,
            ["facts_passed"] = factsPassed,
            ["facts_total"] = factsTotal,
            ["deterministic_grounding_score"] = RoundPct(deterministicScore),
            ["semantic_quality_score"] = RoundPct(LocalSemanticQuality(expected, response)),
            ["semantic_quality_provider"] = "local-token-overlap",
            ["assertion_score"] = RoundPct(assertionScore),
            ["fact_score"] = RoundPct(factScore),
            ["status"] = status,
            ["has_citation"] = HasCitation(response),
            ["has_no_result_language"] = responseNoResult,
            ["expected_no_result"] = expectedNoResult,
            ["failed_checks"] = assertionResults.Where(r => !(bool)r["passed"]!).Select(r => r["value"]).ToList(),
        };
    }

    private static Dictionary<string, object?> Summarize(ScoreAgentConfig agent, List<Dictionary<string, object?>> scores)
    {
        var deterministic = scores.Select(s => ToDouble(s["deterministic_grounding_score"])).ToList();
        var semantic = scores.Select(s => ToDouble(s["semantic_quality_score"])).ToList();
        var totalA = Sum(scores, "assertions_total");
        var passedA = Sum(scores, "assertions_passed");
        var totalF = Sum(scores, "facts_total");
        var passedF = Sum(scores, "facts_passed");
        return new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["agent"] = new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["key"] = agent.Key,
                ["name"] = agent.Name,
                ["connector_id"] = agent.ConnectorId,
            },
            ["prompt_count"] = scores.Count,
            ["average_deterministic_grounding_score"] = Average(deterministic),
            ["average_semantic_quality_score"] = Average(semantic),
            ["assertion_pass_rate"] = totalA > 0 ? RoundPct((double)passedA / totalA) : 100.0,
            ["assertions_passed"] = passedA,
            ["assertions_total"] = totalA,
            ["fact_pass_rate"] = totalF > 0 ? RoundPct((double)passedF / totalF) : 100.0,
            ["facts_passed"] = passedF,
            ["facts_total"] = totalF,
            ["pass_count"] = scores.Count(s => (string?)s["status"] == "pass"),
            ["partial_count"] = scores.Count(s => (string?)s["status"] == "partial"),
            ["fail_count"] = scores.Count(s => (string?)s["status"] == "fail"),
            ["citation_count"] = scores.Count(s => (bool)s["has_citation"]!),
        };
    }

    private static Dictionary<string, object?> SummarizeCategories(List<Dictionary<string, object?>> scores)
    {
        var categories = scores.Select(CategoryOf).Distinct(StringComparer.Ordinal).OrderBy(c => c, StringComparer.Ordinal).ToList();
        var out_ = new Dictionary<string, object?>(StringComparer.Ordinal);
        foreach (var c in categories)
        {
            var items = scores.Where(s => CategoryOf(s) == c).ToList();
            out_[c] = new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["count"] = items.Count,
                ["average_deterministic_grounding_score"] = Average(items.Select(i => ToDouble(i["deterministic_grounding_score"])).ToList()),
                ["average_semantic_quality_score"] = Average(items.Select(i => ToDouble(i["semantic_quality_score"])).ToList()),
            };
        }
        return out_;
    }

    private static string CategoryOf(Dictionary<string, object?> s)
    {
        var v = s["category"]?.ToString();
        return string.IsNullOrEmpty(v) ? "uncategorized" : v;
    }

    private static string RenderMarkdown(Dictionary<string, object> payload, IReadOnlyList<ScoreAgentConfig> agents)
    {
        var lines = new List<string>
        {
            "# Agent Response Scoring",
            string.Empty,
            "Scores include deterministic grounding and a separate semantic quality score.",
            string.Empty,
            "| Agent | Connector | Avg grounding | Avg semantic quality | Assertions passed | Fact pass rate | Pass | Partial | Fail |",
            "|---|---|---:|---:|---:|---:|---:|---:|---:|",
        };
        var agentsPayload = (Dictionary<string, object>)payload["agents"];
        foreach (var agent in agents)
        {
            if (!agentsPayload.TryGetValue(agent.Key, out var raw)) continue;
            var summary = (Dictionary<string, object?>)((Dictionary<string, object>)raw)["summary"]!;
            lines.Add($"| {agent.Name} | `{agent.ConnectorId}` | {summary["average_deterministic_grounding_score"]}% | " +
                $"{summary["average_semantic_quality_score"]}% | {summary["assertions_passed"]}/{summary["assertions_total"]} | " +
                $"{summary["fact_pass_rate"]}% | {summary["pass_count"]} | {summary["partial_count"]} | {summary["fail_count"]} |");
        }
        return string.Join("\n", lines) + "\n";
    }

    // -- TEXT NORMALIZATION & MATCHING ---------------------------------

    /// <summary>NFKD + strip nonspacing marks + lowercase. Mirrors TS
    /// <c>value.normalize('NFKD').replace(/\p{Diacritic}/gu,'').toLowerCase()</c>.</summary>
    public static string FoldText(string value)
    {
        if (string.IsNullOrEmpty(value)) return string.Empty;
        var nfkd = value.Normalize(NormalizationForm.FormKD);
        var sb = new StringBuilder(nfkd.Length);
        foreach (var ch in nfkd)
        {
            if (CharUnicodeInfo.GetUnicodeCategory(ch) == UnicodeCategory.NonSpacingMark) continue;
            sb.Append(ch);
        }
        return sb.ToString().ToLowerInvariant();
    }

    /// <summary>Folded + all non-alphanumerics removed.</summary>
    public static string Compact(string value) =>
        NonAlnumRegex.Replace(FoldText(value), string.Empty);

    private static readonly Regex NonAlnumRegex = new("[^a-z0-9]+", RegexOptions.Compiled);

    public static bool ContainsValue(string response, string expected, bool wholeWord)
    {
        if (string.IsNullOrEmpty(expected)) return true;
        var foldedResp = FoldText(response);
        var foldedExp = FoldText(expected);
        if (wholeWord || IsShortCode(expected))
        {
            var pattern = $"(?<![a-z0-9]){Regex.Escape(foldedExp)}(?![a-z0-9])";
            if (Regex.IsMatch(foldedResp, pattern)) return true;
        }
        return foldedResp.Contains(foldedExp, StringComparison.Ordinal)
            || Compact(response).Contains(Compact(expected), StringComparison.Ordinal);
    }

    public static double LocalSemanticQuality(string expected, string actual)
    {
        var e = TokenSet(expected);
        var a = TokenSet(actual);
        if (e.Count == 0) return a.Count == 0 ? 1.0 : 0.0;
        if (a.Count == 0) return 0.0;
        var overlap = e.Count(t => a.Contains(t));
        var precision = (double)overlap / a.Count;
        var recall = (double)overlap / e.Count;
        return precision + recall == 0 ? 0 : (2 * precision * recall) / (precision + recall);
    }

    private static HashSet<string> TokenSet(string value)
    {
        var folded = FoldText(value);
        var tokens = Regex.Split(folded, "[^a-z0-9.]+");
        return [.. tokens.Where(t => t.Length > 1)];
    }

    public static bool IsShortCode(string value)
    {
        if (string.IsNullOrEmpty(value)) return false;
        var stripped = Regex.Replace(value, "[^A-Za-z0-9]", string.Empty);
        return stripped.Length > 1 && stripped.Length <= 3
            && stripped.Equals(stripped.ToUpperInvariant(), StringComparison.Ordinal);
    }

    public static bool HasNoResultLanguage(string value)
    {
        var folded = FoldText(value);
        return NoResultPatterns.Any(p => folded.Contains(p, StringComparison.Ordinal));
    }

    public static bool HasCitation(string value)
    {
        if (string.IsNullOrEmpty(value)) return false;
        return value.Contains("\ue200cite", StringComparison.Ordinal)
            || value.Contains("cite", StringComparison.Ordinal)
            || Regex.IsMatch(value, @"\[\^\d+\^\]");
    }

    private static IEnumerable<string> NormalizeFacts(IEnumerable<string> facts) =>
        facts.Where(f => f is not null && !string.IsNullOrWhiteSpace(f));

    public static (string Key, string Value) ParseFact(string fact)
    {
        var i = fact.IndexOf('=');
        return i < 0 ? (string.Empty, fact.Trim()) : (fact[..i].Trim(), fact[(i + 1)..].Trim());
    }

    // -- CSV PARSER (matches TS parseCsv) ------------------------------

    public static List<Dictionary<string, string>> ReadCsv(string filePath)
    {
        var rows = ParseCsv(File.ReadAllText(filePath));
        if (rows.Count == 0) return [];
        var header = rows[0];
        var data = new List<Dictionary<string, string>>(rows.Count - 1);
        for (var i = 1; i < rows.Count; i++)
        {
            var dict = new Dictionary<string, string>(StringComparer.Ordinal);
            for (var j = 0; j < header.Count; j++)
            {
                dict[header[j]] = j < rows[i].Count ? rows[i][j] : string.Empty;
            }
            data.Add(dict);
        }
        return data;
    }

    public static List<List<string>> ParseCsv(string content)
    {
        var rows = new List<List<string>>();
        var row = new List<string>();
        var cell = new StringBuilder();
        var inQuotes = false;
        var text = StripBom(content);
        for (var i = 0; i < text.Length; i++)
        {
            var ch = text[i];
            if (ch == '"')
            {
                if (inQuotes && i + 1 < text.Length && text[i + 1] == '"')
                {
                    cell.Append('"');
                    i++;
                }
                else
                {
                    inQuotes = !inQuotes;
                }
            }
            else if (ch == ',' && !inQuotes)
            {
                row.Add(cell.ToString());
                cell.Clear();
            }
            else if ((ch == '\n' || ch == '\r') && !inQuotes)
            {
                if (ch == '\r' && i + 1 < text.Length && text[i + 1] == '\n') i++;
                row.Add(cell.ToString());
                rows.Add(row);
                row = [];
                cell.Clear();
            }
            else
            {
                cell.Append(ch);
            }
        }
        if (cell.Length > 0 || row.Count > 0)
        {
            row.Add(cell.ToString());
            rows.Add(row);
        }
        return rows.Where(cells => cells.Any(v => !string.IsNullOrWhiteSpace(v))).ToList();
    }

    private static string StripBom(string s) =>
        s.Length > 0 && s[0] == '\uFEFF' ? s[1..] : s;

    // -- ARITHMETIC ----------------------------------------------------

    /// <summary>JS Math.round(value * 1000) / 10. Half-away-from-zero per
    /// ECMAScript spec. We use AwayFromZero to match.</summary>
    public static double RoundPct(double value) =>
        Ccw.Core.Util.JsMath.Round(value * 1000) / 10;

    public static double Average(IReadOnlyList<double> values) =>
        values.Count == 0
            ? 0
            : Ccw.Core.Util.JsMath.Round((values.Sum() / values.Count) * 10) / 10;

    private static int Sum(List<Dictionary<string, object?>> values, string key) =>
        values.Sum(v => ToInt(v[key]));

    private static double ToDouble(object? v) => v switch
    {
        null => 0,
        double d => d,
        int i => i,
        long l => l,
        _ => double.TryParse(v.ToString(), NumberStyles.Any, CultureInfo.InvariantCulture, out var d) ? d : 0,
    };

    private static int ToInt(object? v) => v switch
    {
        null => 0,
        int i => i,
        long l => (int)l,
        double d => (int)d,
        _ => int.TryParse(v.ToString(), NumberStyles.Any, CultureInfo.InvariantCulture, out var n) ? n : 0,
    };
}
